### PR TITLE
[fix][broker] Fix compaction subscription delete by inactive subscription check.

### LINKED ISSUE #20982: [Bug] Compaction cursor deleted by inactive subscription check
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

master branch.

### Minimal reproduce step

1. enable compaction.
2. enable SubscriptionExpiration set to 1 minute.

### What did you expect to see?

__compaction subscription should not be deleted.

### What did you see instead?

2023-08-12T13:56:30,821 - INFO  - [main:PersistentTopic@2752] - [persistent://prop/test/testCompactionCursorShouldNotDelete][__compaction] The subscription was deleted due to expiration with last active [1691819767437]

### Anything else?

no

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
Fixes #20982

### Motivation
do not delete compaction cursor when enable subscription inactive check.

### Modifications

if subscription is compaction subscription skip check inactive check.

### Verifying this change

add unit test to check this. see `testCheckInactiveSubscriptionsShouldNotDeleteCompactionCursor`

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

**Comment by lifepuzzlefun** [comment]
@coderzc @gaoran10 please take a look : -)

**Review by coderzc — APPROVED** [review]

**Review by gaoran10 — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java [hunk 1]
```diff
@@ -1218,6 +1220,56 @@ public void testConcurrentLoadTopicExceedLimitShouldNotBeAutoCreated() throws Ex
...
+
+        compactionFuture.get();
+
+        Thread.sleep(2000);
+
+        // this operation is async delete
+        topic.checkInactiveSubscriptionsWithExpirationTime(1000);
+
+        // wait for async delete finish
+        Thread.sleep(2000);
```
Comment (Technoboy-) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/BrokerServiceTest.java:1267]: Better to use `Awaitility`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 2]
```diff
@@ -2747,6 +2739,21 @@ public void checkInactiveSubscriptions() {
         }
     }
 
+    @VisibleForTesting
+    public void checkInactiveSubscriptionsWithExpirationTime(long expirationTimeMillis) {
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2743]: I don't think we should make it to be `PUBLIC`

Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2743]: +1

**Comment by lifepuzzlefun** [comment]
sorry for the late reply. the comments has been applied. please take a look again : -) @mattisonchao @Technoboy-

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 3]
```diff
@@ -247,6 +248,12 @@ protected TopicStatsHelper initialValue() {
 
     // Record the last time a data message (ie: not an internal Pulsar marker) is published on the topic
     private volatile long lastDataMessagePublishedTimestamp = 0;
+
+    // Record the total expired subscription, if enable subscription expiration feature
+    private static final AtomicLongFieldUpdater<PersistentTopic> EXPIRED_SUBSCRIPTION_NUMBERS_UPDATER =
+            AtomicLongFieldUpdater.newUpdater(PersistentTopic.class, "expiredSubscriptionNumbers");
+    private volatile long expiredSubscriptionNumbers = 0L;
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:255]: Seems it's no need to add this

Comment (lifepuzzlefun) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: agree, it is added only for unit test. now i figure out another way to check the logic. PTAL again

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**